src/utils/BucketUtils.ts

```diff
@@ -0,0 +1,384 @@
+      key: '',
+      label: '',
+      datasets: datasets,
+      datasetIds: datasetIds,
+      dataUse: dataUseSummary,
+      dataUses: [...dataUseSummary?.primary || [], ...dataUseSummary?.secondary || []],
+      elections: [],
+      votes: [],
+      matchResults: [],
+      dacs: dacs,
+    }
+    buckets.push(bucket)
+  }
+
+  // The following steps are all bucket-centric, so we can process them in a single loop
+  // Steps 2-6
+  buckets.forEach((b: Bucket) => {
+    // Step 2: Find match results for each dataset in bucket
+    matchData.forEach((m: MatchResult) => {
+      b.datasets.forEach((dataset: Dataset) => {
+        if (dataset.datasetIdentifier.toLowerCase() === m.consent.toLowerCase()) {
+          b.matchResults.push(m)
+        }
+      })
+    })
+
+    // Step 3: Populate elections for datasets in this bucket
+    b.elections = findElectionsForDatasets(recentDar, b.datasetIds)
+
+    // Step 4: Populate votes for each bucket
+    b.votes.push(processVotesForBucket(b.elections))
+
+    // Step 5: Generate bucket key and label
+    b.key = 'bucket-' + b.datasetIds.join('-')
+    b.label = b.dataUses?.map((du: DataUseTerm) => du.code).join(', ') || 'Undefined Data Use'
+
+    // Step 6: Coalesce match results into a single result per bucket
+    b.algorithmResult = calculateAlgorithmResultForBucket(b)
+  })
+
+  // Step 7: Populate RUS Vote bucket with RP votes and move to the top of the bucket array
+  const rpVotes = createRpVoteStructureFromBuckets(buckets)
+  buckets.unshift({
+    isRP: true,
+    key: 'RUS Vote',
+    votes: rpVotes,
+    label: '',
+    datasets: [],
+    datasetIds: [],
+    dataUses: [],
+    elections: [],
+    matchResults: [],
+  })
+
+  return buckets
+}
+
+/**
+ * Find all elections (in a dar) with a dataset id in the provided list of dataset ids
+ * @private
+ * @param dar
+ * @param datasetIds
+ * @returns {Election[]}
+ */
+const findElectionsForDatasets = (dar: DataAccessRequest, datasetIds: number[]): Election[] => {
+  return dar.elections
+    ? Object.values(dar.elections)
+        .filter((e: Election) => datasetIds.includes(e.datasetId))
+    : []
+}
+
+/**
+ * Optionally filter a list of collection datasets by the dac ids provided.
+ * @private
+ * @param dacIds List of dac ids. Can be empty
+ * @param datasets List of datasets to filter
+ * @returns {Dataset[]}
+ */
+const filterDatasetsByDACs = (dacIds: number[], datasets: Dataset[]): Dataset[] => {
+  return isEmpty(dacIds)
+    ? datasets
+    : filter(
+        (dataset: Dataset) => includes(dataset.dacId)(dacIds),
+      )(datasets)
+}
+
+/**
+ * Generate the summary of algorithm results suitable for display in the UI
+ *
+ * Four potential cases:
+ *  1. No matches
+ *  2. Exactly one match or N matches that are all the same - easy case
+ *  3. Abstain is true and match is false - decision is abstained
+ *  4. N matches - not all the same - very confusing case
+ *
+ * @private
+ * @param bucket
+ * @returns {AlgorithmResult}
```

Comment:
Such docstring elements are often nice in JavaScript, but seem unnecessarily duplicative in TypeScript.